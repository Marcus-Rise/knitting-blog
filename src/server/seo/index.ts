export { SeoModule } from "./seo.module";
export type { ISeoConfigService } from "./config";
export { SEO_CONFIG_SERVICE_PROVIDER } from "./config";
export { SEO_SERVICE_PROVIDER } from "./service";
export type { ISeoService } from "./service";
