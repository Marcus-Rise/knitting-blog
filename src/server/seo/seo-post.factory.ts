class SeoPostFactory {
  static toSitemapRaw({
    slug,
    date,
    hostName,
  }: {
    slug: string;
    date: string;
    hostName: string;
  }): string {
    const saveSlug = encodeURIComponent(slug);
    const url = !!saveSlug ? `${hostName}/${saveSlug}/` : `${hostName}/`;
    const lastEditDate = new Date(date);
    const dateStr = lastEditDate.toJSON();

    return `<url>
      <loc>${url}</loc>
      <lastmod>${dateStr}</lastmod>
      </url>`;
  }
}

export { SeoPostFactory };
