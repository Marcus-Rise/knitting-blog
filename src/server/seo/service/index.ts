export * from "./seo-service.provider";
export type { ISeoService } from "./seo-service.interface";
