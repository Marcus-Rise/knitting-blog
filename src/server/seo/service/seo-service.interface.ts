interface ISeoService {
  generateSitemapXml(hostName: string): Promise<string>;
}

export type { ISeoService };
