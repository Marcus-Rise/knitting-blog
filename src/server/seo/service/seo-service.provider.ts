const SEO_SERVICE = Symbol.for("SEO_SERVICE");

export { SEO_SERVICE };
