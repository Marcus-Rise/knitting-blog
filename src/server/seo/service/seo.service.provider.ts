const SEO_SERVICE_PROVIDER = Symbol("ISeoService");

export { SEO_SERVICE_PROVIDER };
