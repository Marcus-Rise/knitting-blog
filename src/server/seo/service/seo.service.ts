import type { ISeoService } from "./seo.service.interface";
import { inject, injectable } from "inversify";
import type { IPostService } from "../../post";
import { POST_SERVICE_PROVIDER } from "../../post";
import type { ISeoConfigService } from "../config";
import { SEO_CONFIG_SERVICE_PROVIDER } from "../config";

@injectable()
class SeoService implements ISeoService {
  constructor(
    @inject(POST_SERVICE_PROVIDER) private readonly posts: IPostService,
    @inject(SEO_CONFIG_SERVICE_PROVIDER) private readonly config: ISeoConfigService,
  ) {}

  static getDateStr(lastEditDate: Date): string {
    return lastEditDate.toJSON();
  }

  static getUrlStr(url: string): string {
    return encodeURIComponent(url);
  }

  async generateRobotsTxt(hostName: string): Promise<string> {
    const sections: Record<string, string> = !this.config.allowRobots
      ? {
          "User-agent": "*",
          Disallow: "/",
          Sitemap: `${hostName}/sitemap.xml`,
        }
      : {
          "User-agent": "*",
          Allow: "/",
          Sitemap: `${hostName}/sitemap.xml`,
        };

    return Object.keys(sections)
      .map((i) => `${i}: ${sections[i]}`)
      .join("\n");
  }

  async generateSitemap(hostName: string): Promise<string> {
    let buf = `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`;
    await this.posts.load(0);

    const date = this.posts.itemLastDate;
    const url = SeoService.getUrlStr(`${hostName}/`);

    if (!!date) {
      buf += `<url>
    <loc>${url}</loc>
    <lastmod>${SeoService.getDateStr(date)}</lastmod>
    </url>`;
    }

    this.posts.items.forEach((post) => {
      const url = SeoService.getUrlStr(`${hostName}/${post.slug}/`);
      const lastEditDate = new Date(post.date);
      const dateStr = SeoService.getDateStr(lastEditDate);

      buf += `<url>
      <loc>${url}</loc>
      <lastmod>${dateStr}</lastmod>
      </url>`;
    });

    buf += `</urlset>`;

    return buf.toString();
  }
}

export { SeoService };
