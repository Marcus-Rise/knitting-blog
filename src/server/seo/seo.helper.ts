import { inject } from "../ioc";
import type { ISeoService } from "./service";
import { SEO_SERVICE } from "./service";

const getRobotsTxt = (hostName: string, isAllow: boolean): string => {
  const robots: Record<string, string> = isAllow
    ? {
        "User-agent": "*",
        Allow: "/",
        Sitemap: `https://${hostName}/sitemap.xml`,
      }
    : {
        "User-agent": "*",
        Disallow: "/",
        Sitemap: `https://${hostName}/sitemap.xml`,
      };

  return Object.keys(robots)
    .map((i) => `${i}: ${robots[i]}`)
    .join("\n");
};

const getSitemap = (hostName: string) =>
  inject((seoService: ISeoService) => seoService.generateSitemapXml(hostName), [SEO_SERVICE]);

export { getRobotsTxt, getSitemap };
