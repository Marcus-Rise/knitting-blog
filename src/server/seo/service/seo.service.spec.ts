import "reflect-metadata";
import { SeoService } from "./seo.service";
import { mock } from "jest-mock-extended";
import type { IPost, IPostService } from "../../post";
import type { ISeoConfigService } from "../config";

describe("SeoService", () => {
  describe("generateRobotsTxt", () => {
    test("robots are not allowed", async () => {
      const service = new SeoService(
        mock<IPostService>(),
        mock<ISeoConfigService>({
          allowRobots: false,
        }),
      );

      const data = await service.generateRobotsTxt("hostname");

      expect(data.length).toBeGreaterThan(0);
      expect(data).toMatchSnapshot();
    });
    test("robots are allowed", async () => {
      const service = new SeoService(
        mock<IPostService>(),
        mock<ISeoConfigService>({ allowRobots: true }),
      );

      const data = await service.generateRobotsTxt("hostname");

      expect(data.length).toBeGreaterThan(0);
      expect(data).toMatchSnapshot();
    });
  });

  test("generateSitemap", async () => {
    const service = new SeoService(
      mock<IPostService>({
        items: [
          mock<IPost>({
            slug: "post1",
            date: new Date(2020, 12, 12).toJSON(),
          }),
          mock<IPost>({
            slug: "post2",
            date: new Date(2020, 12, 13).toJSON(),
          }),
        ],
      }),
      mock<ISeoConfigService>(),
    );

    const data = await service.generateSitemap("hostname");

    expect(data.length).toBeGreaterThan(0);
    expect(data).toMatchSnapshot();
  });
});
