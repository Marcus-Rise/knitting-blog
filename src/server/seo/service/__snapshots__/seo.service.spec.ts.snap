// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeoService generateRobotsTxt robots are allowed 1`] = `
"User-agent: *
Allow: /
Sitemap: https://hostname/sitemap.xml"
`;

exports[`SeoService generateRobotsTxt robots are not allowed 1`] = `
"User-agent: *
Disallow: /
Sitemap: https://hostname/sitemap.xml"
`;

exports[`SeoService generateSitemap 1`] = `
"<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url>
      <loc>https://%D0%BD%D0%B0%D0%B4%D1%8F-%D0%B2%D1%8F%D0%B6%D0%B5%D1%82.%D1%80%D1%84/%D0%BC%D1%83%D0%B6%D1%81%D0%BA%D0%BE%D0%B8-%D0%B6%D0%B0%D0%BA%D0%B5%D1%82-%D1%81-%D0%B2%D0%BE%D1%80%D0%BE%D1%82%D0%BD%D0%B8%D0%BA%D0%BE%D0%BC-%D1%81%D1%82%D0%BE%D0%B8%D0%BA%D0%BE%D0%B8/</loc>
      <lastmod>2021-01-11T20:00:00.000Z</lastmod>
      </url><url>
      <loc>https://%D0%BD%D0%B0%D0%B4%D1%8F-%D0%B2%D1%8F%D0%B6%D0%B5%D1%82.%D1%80%D1%84/post2/</loc>
      <lastmod>2021-01-12T20:00:00.000Z</lastmod>
      </url></urlset>"
`;
